Add PhpArrayAdapter to cache metadata

ossinkine opened the pull request:
Closes #1186

cc @ostrolucky @dmaicher @nicolas-grekas 

I see two weak places in this implementation:
1. `DoctrineMetadataCacheWarmer` extends `AbstractPhpFileCacheWarmer` which also extended by `ValidatorCacheWarmer`, `SerializerCacheWarmer` and `AnnotationsCacheWarmer` (everywhere `PhpArrayAdapter` used) but it's marked as internal, so BC breaks possible here. I can avoid extending this class but a lot of code will be copy-pasted.
2. Since `PhpArrayAdapter` requires Symfony cache I wrap Doctrine cache with `DoctrineAdapter`. If Doctrine cache is `DoctrineProvider` which wrap a Symfony cache we get a redudant chain of adapters. For example the chain `PhpArrayAdapter -> DoctrineAdapter -> DoctrineProvider -> RedisAdapter` can be simplified by removing `DoctrineAdapter` and `DoctrineProvider`: `PhpArrayAdapter -> RedisAdapter`. I've tried to create a compiler pass which remove this redundancy but get list of entity managers in compiler pass a little bit difficult. And maybe we can implement this simplification in Symfony framework bundle globally for all such cache chains.

What is your opinion about this?

nicolas-grekas commented:
> `AbstractPhpFileCacheWarmer` is internal

That can be changed, PR welcome on symfony/symfony to discuss it there.

> DoctrineAdapter -> DoctrineProvider

This would be better implemented on the bundle imho, as that would allow releasing the logic in sync with the feature provided here.

ostrolucky commented:
Tests are DI only, very simplistic. Can we have test which asserts that cache is hit when loading metadata and not generated again? Also, can you provide some numbers so everyone can see what's the expected improvement?

ostrolucky commented on DependencyInjection/DoctrineExtension.php:
```diff
@@ -898,4 +906,34 @@ private function createArrayAdapterCachePool(ContainerBuilder $container, string
+        $cacheWarmerDefinition->addTag('kernel.cache_warmer');
+    }
+
+    private function registerMetadataPhpArrayCache(string $entityManagerName, ContainerBuilder $container): void
```
I would personally prefer to not unnecessarily register new service if its only purpose is to inject it to warmer. Service should be instead just inlined, IMHO. Otherwise we are unnecessarily increasing public API surface.

ossinkine replied:
`doctrine.orm.default_metadata_cache` is the alias which refers to the cache service. How can I create inline service and refer the alias to it?

ostrolucky commented on DependencyInjection/Compiler/PhpArrayCachePass.php:
```diff
@@ -0,0 +1,44 @@
+/**
+ * Removes redundant decorators such DoctrineAdapter -> DoctrineProvider -> AdapterInterface.
+ */
+class PhpArrayCachePass implements CompilerPassInterface
```
We would prefer not having to maintain this in bundle forever. Can you please contribute this to Symfony core too? That way you will also get better feedback, as I'm not exactly confident about these internals.

ossinkine replied:
As I understand @nicolas-grekas has another opinion. Since this is just optimization we can merge without it.

ostrolucky replied:
Nicolas didn't mean it doesn't make sense to contribute to Symfony, he said it needs to be merged to doctrine bundle anyways, as otherwise this feature would be available in next minor Symfony version only. But yeah I am also ok with just removing it for now, would make it easier to merge

ostrolucky reviewed (changes requested):
Tests are DI only, very simplistic. Can we have test which asserts that cache is hit when loading metadata and not generated again? Also, can you provide some numbers so everyone can see what's the expected improvement?

Also, we probably need to mark a test as legacy so CI passes.
And of course, needs to be rebased as it has conflicts.

ossinkine commented:
Tests don't passed because AbstractPhpFileCacheWarmer marked as internal, `@group legacy` does not hide the errors

nicolas-grekas commented:
Hmm, I didn't consider this aspect.
I'd be fine removing the annotation on branch 3.4 on Symfony.

ostrolucky commented:
Any chance for finishing this @ossinkine?

ossinkine commented:
@ostrolucky I can remove `PhpArrayCachePass` which optimizes cache adapter chaining right now and later publish PR to Symfony but I don't what tests you requested and I don't how much time it will take

ostrolucky commented:
Sounds good enough to me!

ossinkine commented:
@ostrolucky I've removed the commit with optimizations, rebased the branch from the master and fixed code style issue except the one, I think proposed fix is a bad idea

ostrolucky commented:
How about this. Does that work for you @ossinkine?

stof commented:
> get list of entity managers in compiler pass a little bit difficult.

this could actually be simple. We can add a tag on the services and use that tag in the compiler pass

stof commented on DependencyInjection/Configuration.php:
```diff
@@ -735,7 +741,7 @@ private function getDeprecationMsg(string $message): array
             return [
                 'doctrine/doctrine-bundle',
-                '2.0',
+                '2.2',
```
This change is wrong. If different methods need different versions, you need to pass that as an argument, not changing the version for previous deprecations

ostrolucky replied:
it's already wrong in master, will fix in separate commit

stof commented on Tests/DependencyInjection/DoctrineExtensionTest.php:
```diff
@@ -776,6 +793,7 @@ public function testInvalidCacheConfiguration(): void
      * @param array|string $cacheConfig
      *
      * @dataProvider cacheConfigurationProvider
+     * @group legacy
```
The legacy cases must be split into separate tests. Not everything about cache configuration is legacy.

ostrolucky replied:
i don't this it's worth it

stof replied:
to me it is, for 2 reasons:

- when you remove a deprecated feature, you can remove the associated legacy tests. But here, that would remove the coverage for other cache configurations too
- marking tests covering other (non-deprecated) features as `@legacy` can hide accidental usage of deprecated APIs (which is precisely the reason to make test fail on deprecations in non-legacy tests)

dmaicher commented on DependencyInjection/Configuration.php:
```diff
@@ -688,6 +688,13 @@ private function getOrmCacheDriverNode(string $name): ArrayNodeDefinition
                 ->scalarNode('pool')->end()
             ->end();
 
+        if ($name === 'metadata_cache_driver') {
```
this means we also need to create a new recipe for 2.2?

See https://github.com/symfony/recipes/blob/master/doctrine/doctrine-bundle/1.12/config/packages/prod/doctrine.yaml#L4 (it's used for 2.x currently as well via symlink)

ostrolucky replied:
should yes

ossinkine commented:
@ostrolucky since PHP Array Cache adapter is a decorator for another adapter I'm not sure `metadata_cache_driver` should be depreacated. Or you suppose there can't be cache misses?

ostrolucky commented:
If PHP Array Cache adapter is a decorator for another adapter, what happens right now when `metadata_cache_driver` is not set?

greg0ire commented on DependencyInjection/Configuration.php:
```diff
@@ -688,6 +688,13 @@ private function getOrmCacheDriverNode(string $name): ArrayNodeDefinition
 
+        if ($name === 'metadata_cache_driver') {
+            $node->setDeprecated(...$this->getDeprecationMsg(
+                'The "metadata_cache_driver" configuration key is deprecated. PHP Array cache is now automatically registered when %kernel.debug% is false.',
```
Why are we deprecating this configuration key? There should be an explanation in the commit message.

ostrolucky replied:
We are deprecating it because it doesn't make sense to specify any other cache than what we automatically configure now. If we are leaving it, then it would make more sense to change flex recipe instead.

greg0ire commented on DependencyInjection/DoctrineExtension.php:
```diff
@@ -927,4 +936,25 @@ private function createArrayAdapterCachePool(ContainerBuilder $container, string
+    {
+        $metadataCacheAlias              = $this->getObjectManagerElementName($entityManagerName . '_metadata_cache');
+        $decoratedMetadataCacheServiceId = (string) $container->getAlias($metadataCacheAlias);
+        $phpArrayCacheDecoratorServiceId = $decoratedMetadataCacheServiceId . '.php_array';
```
Why is there decoration involved?

ostrolucky replied:
PhpArrayAdapter requires it https://github.com/symfony/symfony/blob/57e39b41b94dd9d0beb04c7f88fc0955a5bbe605/src/Symfony/Component/Cache/Adapter/PhpArrayAdapter.php#L48. Also for BC, so current configuration is not completely ignored until we remove it.

greg0ire reviewed (commented):
Please state the obvious. It would be great not to have to read several issues and PRs on different repositories to understand this.

greg0ire reviewed (approved):
I pushed new commits with explanations that might be completely wrong, please review and amend them if necessary.

stephanvierkant commented:
It isn't clear to me what I've to change in my application and I'm not the only one: https://stackoverflow.com/questions/64741823/symfony-4-4-metadata-cache-driver-configuration-key-deprecation-notice

Can someone explain why this change was needed and how we should replace `metadata_cache_driver`? If I understand, I can update the Symfony docs (if needed).

ostrolucky commented:
metadata_cache_driver configuration needs to be removed. Change is needed because defining own metadata_cache_driver is useless from now on.

dmaicher commented on DependencyInjection/DoctrineExtension.php:
```diff
@@ -927,4 +936,25 @@ private function createArrayAdapterCachePool(ContainerBuilder $container, string
+        $phpArrayCacheDecoratorServiceId = $decoratedMetadataCacheServiceId . '.php_array';
+        $phpArrayFile                    = '%kernel.cache_dir%' . sprintf('/doctrine/orm/%s_metadata.php', $entityManagerName);
+
+        $container->register(DoctrineMetadataCacheWarmer::class)
```
@ossinkine actually this is not working properly for multiple entity managers. I just tried it on one of my bigger projects that uses 2 managers.

Since `DoctrineMetadataCacheWarmer::class` is the service id it will only work for the last manager and overwrite any previous definitions
